### PR TITLE
[Misc] Add uninitialized params tracking for `AutoWeightsLoader`

### PR DESCRIPTION
**Pull request by Isotr0py**
Currently, model have uninitialized parameters won't raise any error even a warning message. This makes some bugs like https://github.com/vllm-project/vllm/pull/10297#discussion_r1840989950 and https://github.com/vllm-project/vllm/pull/9422#issuecomment-2418746915 caused by incorrect weights loading/model initialization difficult to be found.
- Add feature to track uninitialized weights for `AutoWeightsLoader`.

### CODE REVIEW

File: vllm/model_executor/models/utils.py [hunk 1]
```diff
@@ -172,8 +173,9 @@ def _load_module(
         if module != self.module:
             module_load_weights = getattr(module, "load_weights", None)
             if callable(module_load_weights):
-                module_load_weights(weights)
-                return
+                loaded_params = module_load_weights(weights)
+                yield from map(lambda x: self._get_qualname(base_prefix, x),
+                               loaded_params)
```
Comment (Isotr0py) [inline vllm/model_executor/models/utils.py:178]: I also modified `load_weights` in most of models to return loaded params' names for alignment and compatibility, even though they aren't used with `AutoWeightsLoader`.

File: vllm/model_executor/models/utils.py [hunk 2]
```diff
@@ -222,11 +224,15 @@ def load_weights(
...
+    ) -> Set[str]:
         if mapper is not None:
             weights = mapper.apply(weights)
 
-        autoloaded_weights = list(self._load_module("", self.module, weights))
+        autoloaded_weights = set(self._load_module("", self.module, weights))
+        if (weights_not_loaded := self.weights_to_load - autoloaded_weights):
+            logger.warning(
+                "Following weights were not initialized from checkpoint: %s",
+                weights_not_loaded)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/utils.py:235]: Can we have a setting (like `strict=True`) to raise an error when any weights are not loaded? Then we can robustly test the weight loading of all models in `tests/models/test_initialization.py`.

**Review by DarkLight1337 — APPROVED** [review]
Thanks for the QoL improvement!

**Comment by DarkLight1337** [comment]
Hmm. not sure whether we're being too strict or is the weight loader buggy?

**Comment by Isotr0py** [comment]
Hmmm, seems that `strict=True` is too strict for sharded safetensors, because in that case the weights of a module will be sharded into multiple files, and it will trigger strict check then raise error while the module is partially loaded from one of the shared files...

Perhaps we should add the loading check on top of `DefaultModelLoader` in `loader.py` instead of `AutoWeightsLoader`.

**Comment by DarkLight1337** [comment]
Agreed, let's move the weight loading check to a higher level.

File: vllm/model_executor/model_loader/loader.py [hunk 3]
```diff
@@ -331,7 +331,16 @@ def load_model(self, vllm_config: VllmConfig) -> nn.Module:
             with target_device:
                 model = _initialize_model(vllm_config=vllm_config)
 
-            model.load_weights(self._get_all_weights(model_config, model))
+            weights_to_load = {name for name, _ in model.named_parameters()}
+            loaded_weights = model.load_weights(
+                self._get_all_weights(model_config, model))
+            # We only enable strict check for non-quantiized models
+            # that have loaded weights tracking currently.
+            if model_config.quantization is None and loaded_weights is not None:
```
Comment (DarkLight1337) [inline vllm/model_executor/model_loader/loader.py:339]: Shouldn't we check `weights_not_loaded` here?

File: vllm/model_executor/model_loader/loader.py [hunk 4]
```diff
@@ -331,7 +331,16 @@ def load_model(self, vllm_config: VllmConfig) -> nn.Module:
...
+                weights_not_loaded = weights_to_load - loaded_weights
```
Comment (DarkLight1337) [inline vllm/model_executor/model_loader/loader.py:340]:
```suggestion
            if model_config.quantization is None and loaded_weights is not None:
                weights_not_loaded = weights_to_load - loaded_weights
```
```suggestion
            weights_not_loaded = weights_to_load - loaded_weights
            if model_config.quantization is None and weights_not_loaded:
                
```

**Comment by DarkLight1337** [comment]
Since we may skip loading some layers in vision models / PP case, we should update the check to error out when weights inside the initialized model fail to be loaded, rather than when some weights from the checkpoint aren't loaded.

**Comment by DarkLight1337** [comment]
I found that this PR causes speculative tests to fail on main. PTAL @Isotr0py

**Pull request merged**